pytorch_lightning/loggers/wandb.py

```diff
@@ -209,6 +207,18 @@ def version(self) -> Optional[str]:
 
     @rank_zero_only
     def finalize(self, status: str) -> None:
-        # upload all checkpoints from saving dir
+        # save checkpoints as artifacts
         if self._log_model:
-            wandb.save(os.path.join(self.save_dir, "*.ckpt"))
+            # use run name and ensure it's a valid Artifact name
+            artifact_name = re.sub(r"[^a-zA-Z0-9_\.\-]", "", self.experiment.name)
+            # gather interesting metadata
+            metadata = {
+                k: v
+                for k, v in dict(self.experiment.summary).items()
+                if isinstance(v, numbers.Number) and not k.startswith("_")
+            }
+            # TODO: see if we can also log data from `trainer.checkpoint_callback` (best_model_path, etc) 
+            artifact = wandb.Artifact(name=f"run-{artifact_name}", type="model", metadata=metadata)
+            # TODO: we need access to `trainer.checkpoint_callback.dirpath`
```

Comment:
@awaelchli  Do you know how I could access `train.checkpoint_callback.dirpath`?
I just need to know where the models are located. If I can access the entire callback, I could actually also log interesting values (best_model_path, etc).

Reply:
hmm no, the trainer is unfortunately nowhere accessible in the logger. A possible solution is to pass the trainer object to the finalize method. This would have to be updated in a few other places and loggers though.

Reply:
What about passing it at init so loggers can access all items they need at that point (or save a reference)?

Reply:
Not in love with this idea tbh. I believe our wrappers around the 3rd party loggers should also be usable without a trainer. 
How about a connect method: At the appropriate time, say in fit, trainer calls
`self.logger.connect(self)` passes itself or maybe also the model, like `self.logger.connect(self, model)`. 
The logger can then read required attributes from trainer.

Reply:
Yes that would work. I imagine it will iterate over all the loggers (or also the callbacks)?

What about just passing a reference to ModelCheckpoint (if the objective is to keep the Trainer separate from loggers) which defaults to None?

Reply:
@borisdayma I think a connect method works best (would probably be called by `LoggerConnector`).

If you give it a go, it should be fairly easy to implement the baseclass method for it to do actually nothing (just for signature consistency).

I would recommend to only pass in the trainer and preferably only save a `weakref` or proxy to it, since loggers should not have a reference that keeps it alive :)

Depending on future other tasks the `ModelCheckpoint may not be sufficient and for other parts we also pass the trainer there (just not upon init so it's optional).

Reply:
So I'm looking at a few places based on your suggestions where this connection can be made. We could update `LoggerConnector.configure_logger` (which attaches logger to the trainer):
1. we just add the reverse operation, attaching the trainer to the logger (as a weakref or proxy)
2. we just add a call to `logger.connect` function which does nothing by default

Does one of these methods sound reasonable?

Reply:
I would go with option 2